### PR TITLE
OpenIdConnectHandler Redirect to Identity Provider Need Consider Ajax request to avoid CORS error

### PR DESCRIPTION
**Pull request by creativebrother**
The issue is raised in Microsoft.Identity.Web project, where a dynamic consent on controller action called by ajax method cause user agent CORS error. A seperate PR https://github.com/AzureAD/microsoft-identity-web/pull/665 is suggested by the team.

Suggest enhancement
(1) in Microsoft.Identity.Web project, update AuthorizeForScopesAttribute to extract a return url from ajax custom header. This return url is for **user agent to return to after the challenge is satisfied**.
(2) in aspnetcore project, for OpenIdConnentHandler,  after the **redirect url for Identity Provider** is calculated, for ajax request, the handler will return a 401 with the redirect url in the header location.

Added a test project in a seperate PR https://github.com/AzureAD/microsoft-identity-web/pull/665 for Microsoft.Identity.Web to test modified AuthorizeForScopesAttribute to handle Ajax call. It depends on a locally built microsoft.aspnetcore.authentication.openidconnect dev nupkg within which associate change on ajax call redirection is handled differently (OpenIdConnectHandler need to be updated at Asp.net Core side).

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=26689) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=26689) before we can accept your contribution.<br/><br/>:x: creativebrother [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=26689)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=26689) it.</sub>

**Comment by javiercn** [comment]
Is the issue here that some JS is calling and API and emitting a challenge is causing a redirect + CORS error?

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs [hunk 1]
```diff
@@ -462,8 +462,13 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
...
                 }
-
-                Response.Redirect(redirectUri);
+                if (IsAjaxRequest(Context.Request))
+                {
+                    Response.Headers["Location"] = redirectUri;
+                    Response.StatusCode = 401;
+                }
+                else
+                    Response.Redirect(redirectUri);
```
Comment (javiercn) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:471]: Brackets

File: .gitignore [hunk 2]
```diff
@@ -43,3 +43,4 @@ msbuild.ProjectImports.zip
 StyleCop.Cache
 UpgradeLog.htm
 .idea
+/src/Framework/App.Ref/src/PackageOverrides.txt
```
Comment (Tratcher) [inline .gitignore:46]: ?

**Review by Tratcher — COMMENTED** [review]
I assume the other redirect mode (`AuthenticationMethod = OpenIdConnectRedirectBehavior.FormPost`) is even worse in this scenario?

**Comment by creativebrother** [comment]
> Is the issue here that some JS is calling and API and emitting a challenge is causing a redirect + CORS error?

Yes, due to nature of the "transparent redirect" that the JS has no control hence CORS error.

File: .gitignore (hunk 2 above)
Comment (creativebrother) [reply .gitignore:46]: That is emitted by visual studio preview. Guess we do not need that.

**Comment by javiercn** [comment]
> Yes, due to nature of the "transparent redirect" that the JS has no control hence CORS error.

You can specify `redirect: manual` in the fetch options to avoid this. I'm not sure the `OpenIdConnectHandler` should accommodate for that. I would expect JS calling APIs to use JWTs instead of (I assume) cookies.

@Tratcher do you have any thoughts?

**Comment by Tratcher** [comment]
We have prior art for this in cookie auth, not to address CORS specifically, but SPAs in general. There we did it for all redirects, not just the challenge.
https://github.com/dotnet/aspnetcore/blob/ed4069c2634ce787025f2781e99328a48dd92802/src/Security/Authentication/Cookies/src/CookieAuthenticationEvents.cs#L41-L49

You would have the same CORS problem with remote signouts, no?
https://github.com/dotnet/aspnetcore/blob/dac42fb8d0a0008dd6bf4ac795fa856ee2e373f8/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L264

The same workaround should be possible via the RedirectToIdentityProvider event, though you'd have to copy a few more lines of code from challenge. That would still be recommended over forking the whole package.

@blowdart If we add this as a first class feature in OIDC should we do it for all the redirects here, and should we also do it for OAuth?

**Comment by blowdart** [comment]
If you're going to make it first class I think you, and cookies needs to make IsAjaxRequest() overridable, so devs can tweak it.

**Comment by Tratcher** [comment]
> If you're going to make it first class I think you, and cookies needs to make IsAjaxRequest() overridable, so devs can tweak it.

At that point you go back to the event (RedirectToIdentityProvider).

**Comment by creativebrother** [comment]
> We have prior art for this in cookie auth, not to address CORS specifically, but SPAs in general. There we did it for all redirects, not just the challenge.
> https://github.com/dotnet/aspnetcore/blob/ed4069c2634ce787025f2781e99328a48dd92802/src/Security/Authentication/Cookies/src/CookieAuthenticationEvents.cs#L41-L49
> 
> You would have the same CORS problem with remote signouts, no?
> https://github.com/dotnet/aspnetcore/blob/dac42fb8d0a0008dd6bf4ac795fa856ee2e373f8/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L264
> 
> The same workaround should be possible via the RedirectToIdentityProvider event, though you'd have to copy a few more lines of code from challenge. That would still be recommended over forking the whole package.
> 
> @blowdart If we add this as a first class feature in OIDC should we do it for all the redirects here, and should we also do it for OAuth?

@Tratcher That's what I did to solve my problem currently via RedirectToIdentityProvider event. But it takes a while to figure that out for me until after you hinted about the cookie auth. I was using the Microsoft.Identity.Web and it has a version of RedirectToIdentityProvider in its extension method we have to include in to make the whole thing work. 
That's why I suggested some enhancement from OpenId side..
It would be great to help the average developers to save time and headache for this pretty confusing issue.

**Comment by Tratcher** [comment]
@javiercn's right that JWTs are probably better in this situation. How does the incremental consent work for JWTs?

**Comment by creativebrother** [comment]
> @javiercn's right that JWTs are probably better in this situation. How does the incremental consent work for JWTs?

@javiercn and @Tratcher 
My case is JS call Web App instead of API(a confidential client app, not an SPA), JS does not handle access token directly, just for clarification.  Web app need access token (JWT) to call downstream API and tokens cached within web app. When cached token does not satisfy the scopes for the resources, an incremental scope challenge is raised.
The incremental consent is "Incremental" such that the JS does not anticipate ahead of time what token it will acquire for the scopes so the challenge itself is essentially for such a token be granted by user (user consent). 
I do not think @javiercn's suggestion apply here?
Thanks for sharing your super awesome opinions here! Love it. Thanks to @javiercn  and @Tratcher!

**Comment by blowdart** [comment]
Can you explain JS "call web app" please? It's really hard to picture this scenario. Where would a token come from here, where is it stored? How is it updated?

Do you have a minimal example app we can look at?

**Comment by Tratcher** [comment]
We decided not to merge this PR. This is a non-standard flow that is mixing concerns for webapps and webapis in ways OIDC wasn't designed for. You can use the existing extensibility points to get the behavior you're after, but it's not a pattern we want to endorse.

**Pull request closed without merging**